## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from tensorflow.keras.layers import Input, Dense, concatenate

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
attrition_df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [3]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [4]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]

y_df.head()


Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [5]:
# Create a list of at least 10 column names to use as X data
X_column_list = ['Age', 'BusinessTravel', 'DistanceFromHome',
                 'EnvironmentSatisfaction', 'HourlyRate',
                 'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
                 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
                 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
                 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
                 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
                 'YearsWithCurrManager']


# Create X_df using your selected columns
X_df = attrition_df[X_column_list]

# Show the data types for X_df
X_df.dtypes


Age                          int64
BusinessTravel              object
DistanceFromHome             int64
EnvironmentSatisfaction      int64
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
NumCompaniesWorked           int64
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StockOptionLevel             int64
TotalWorkingYears            int64
TrainingTimesLastYear        int64
WorkLifeBalance              int64
YearsAtCompany               int64
YearsInCurrentRole           int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [6]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)



In [7]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

# See following cells for encoding processes


In [8]:
# Look at the 3 columns with object data types to understand best way to encode.
X_df[['BusinessTravel', 'JobRole', 'OverTime']]

BusinessTravel                    JobRole OverTime
0         Travel_Rarely            Sales Executive      Yes
1     Travel_Frequently         Research Scientist       No
2         Travel_Rarely      Laboratory Technician      Yes
3     Travel_Frequently         Research Scientist      Yes
4         Travel_Rarely      Laboratory Technician       No
...                 ...                        ...      ...
1465  Travel_Frequently      Laboratory Technician       No
1466      Travel_Rarely  Healthcare Representative       No
1467      Travel_Rarely     Manufacturing Director      Yes
1468  Travel_Frequently            Sales Executive       No
1469      Travel_Rarely      Laboratory Technician       No

[1470 rows x 3 columns]

In [9]:
# Check for null values
null_counts = X_df.isnull().sum()
print(null_counts)

Age                         0
BusinessTravel              0
DistanceFromHome            0
EnvironmentSatisfaction     0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
NumCompaniesWorked          0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64


In [10]:
# Create lists for OneHotEncoder and LabelEncoder

enc_columns = ['BusinessTravel', 'OverTime']
LE_columns = ['JobRole']

In [11]:
# Create a OneHotEncoder for BusinessTravel and Overtime columns
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(drop='if_binary', sparse=False)

# encode the columns included in the enc_columns list. Do this for both train and test sets.
X_encoded_array_train = encoder.fit_transform(X_train[enc_columns])
X_encoded_array_test = encoder.fit_transform(X_test[enc_columns])

print(X_encoded_array_train[0:5])
print(X_encoded_array_test[0:5])

[[0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 1. 0.]
 [0. 0. 1. 0.]
 [0. 0. 1. 0.]]
[[0. 0. 1. 0.]
 [0. 0. 1. 1.]
 [0. 0. 1. 1.]
 [0. 0. 1. 0.]
 [0. 0. 1. 0.]]


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [12]:
# use get feature names to get list of columns
encoded_columns = encoder.get_feature_names_out(enc_columns)

# convert encoded arrays into dataframes
X_enc_train = pd.DataFrame(X_encoded_array_train, columns=encoded_columns)
X_enc_test = pd.DataFrame(X_encoded_array_test, columns=encoded_columns)

X_enc_test


BusinessTravel_Non-Travel  BusinessTravel_Travel_Frequently  \
0                          0.0                               0.0   
1                          0.0                               0.0   
2                          0.0                               0.0   
3                          0.0                               0.0   
4                          0.0                               0.0   
..                         ...                               ...   
363                        0.0                               0.0   
364                        0.0                               1.0   
365                        0.0                               0.0   
366                        0.0                               1.0   
367                        0.0                               0.0   

     BusinessTravel_Travel_Rarely  OverTime_Yes  
0                             1.0           0.0  
1                             1.0           1.0  
2                             1.0           1.0  
3                             1.0           0.0  
4                             1.0           0.0  
..                            ...           ...  
363                           1.0           0.0  
364                           0.0           0.0  
365                           1.0           1.0  
366                           0.0           1.0  
367                           1.0           0.0  

[368 rows x 4 columns]

In [13]:
# Create Label Encoder for JobRole column. Using Label encoder because there are 9 different possible responses and want to avoid adding 9 additional columns.

from sklearn.preprocessing import LabelEncoder

LE = LabelEncoder()

# Create copy of X_df to use Label Encoder. Will then concatenate that LE dataframe with the OneHotEncoder df. Will need to remember to drop the original un-encoded columns

X_train_label_encoded = X_train.copy()
X_test_label_encoded = X_test.copy()

for column in LE_columns:
  X_train_label_encoded[column] = LE.fit_transform(X_train_label_encoded[column])
  X_test_label_encoded[column] = LE.fit_transform(X_test_label_encoded[column])

X_train_label_encoded.reset_index(drop=True, inplace=True)
X_test_label_encoded.reset_index(drop=True, inplace=True)

X_test_label_encoded




Age     BusinessTravel  DistanceFromHome  EnvironmentSatisfaction  \
0     33      Travel_Rarely                13                        2   
1     40      Travel_Rarely                10                        4   
2     30      Travel_Rarely                 1                        4   
3     35      Travel_Rarely                22                        2   
4     36      Travel_Rarely                26                        1   
..   ...                ...               ...                      ...   
363   46      Travel_Rarely                 9                        1   
364   33  Travel_Frequently                29                        1   
365   42      Travel_Rarely                24                        4   
366   27  Travel_Frequently                 5                        4   
367   29      Travel_Rarely                 6                        2   

     HourlyRate  JobInvolvement  JobLevel  JobRole  JobSatisfaction  \
0            53               3         1        6                4   
1            46               4         1        2                3   
2            55               2         2        2                4   
3            71               4         3        3                2   
4            80               3         2        0                3   
..          ...             ...       ...      ...              ...   
363          52               3         3        7                4   
364          54               2         2        6                3   
365          56               3         3        4                1   
366          53               3         1        2                4   
367          54               3         1        2                4   

     NumCompaniesWorked  ... PerformanceRating  RelationshipSatisfaction  \
0                     3  ...                 3                         1   
1                     3  ...                 3                         3   
2                     1  ...                 3                         3   
3                     7  ...                 3                         2   
4                     6  ...                 3                         2   
..                  ...  ...               ...                       ...   
363                   4  ...                 3                         1   
364                   1  ...                 4                         2   
365                   1  ...                 4                         4   
366                   0  ...                 3                         3   
367                   1  ...                 3                         1   

     StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                   0                  5                      4   
1                   1                 10                      3   
2                   2                 10                      1   
3                   1                 10                      6   
4                   2                 10                      2   
..                ...                ...                    ...   
363                 1                 28                      1   
364                 0                 14                      4   
365                 1                 10                      5   
366                 0                  6                      3   
367                 1                 10                      5   

     WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                  3               3                   2   
1                  3               7                   7   
2                  2              10                   8   
3                  2               7                   7   
4                  2               3                   2   
..               ...             ...                 ...   
363                4               7                   7   
364                3              13                   7   
365  

In [14]:
# Merge OneHotEncoded dataframe to label encoded dataframe and drop original unencoded columns
# Used merge because putting together dataframes with different column names.

X_train_final = pd.merge(X_train_label_encoded, X_enc_train, left_index=True, right_index=True, how='inner')
X_test_final = pd.merge(X_test_label_encoded, X_enc_test, left_index=True, right_index=True, how='inner')


X_train_final.reset_index(drop=True, inplace=True)
X_test_final.reset_index(drop=True, inplace=True)

X_train_final.drop(['BusinessTravel', 'OverTime'], axis=1, inplace=True)
X_test_final.drop(['BusinessTravel', 'OverTime'], axis=1, inplace=True)

X_test_final

Age  DistanceFromHome  EnvironmentSatisfaction  HourlyRate  \
0     33                13                        2          53   
1     40                10                        4          46   
2     30                 1                        4          55   
3     35                22                        2          71   
4     36                26                        1          80   
..   ...               ...                      ...         ...   
363   46                 9                        1          52   
364   33                29                        1          54   
365   42                24                        4          56   
366   27                 5                        4          53   
367   29                 6                        2          54   

     JobInvolvement  JobLevel  JobRole  JobSatisfaction  NumCompaniesWorked  \
0                 3         1        6                4                   3   
1                 4         1        2                3                   3   
2                 2         2        2                4                   1   
3                 4         3        3                2                   7   
4                 3         2        0                3                   6   
..              ...       ...      ...              ...                 ...   
363               3         3        7                4                   4   
364               2         2        6                3                   1   
365               3         3        4                1                   1   
366               3         1        2                4                   0   
367               3         1        2                4                   1   

     PercentSalaryHike  ...  TrainingTimesLastYear  WorkLifeBalance  \
0                   18  ...                      4                3   
1                   13  ...                      3                3   
2                   12  ...                      1                2   
3                   18  ...                      6                2   
4                   14  ...                      2                2   
..                 ...  ...                    ...              ...   
363                 11  ...                      1                4   
364                 22  ...                      4                3   
365                 21  ...                      5                2   
366                 14  ...                      3                2   
367                 13  ...                      5                3   

     YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0                 3                   2                        0   
1                 7                   7                        1   
2                10                   8                        3   
3                 7                   7                        6   
4                 3                   2                        0   
..              ...                 ...                      ...   
363               7                   7                        4   
364              13                   7                        3   
365              10                   9                        5   
366               5                   4                        0   
367              10                   8                        0   

     YearsWithCurrManager  BusinessTravel_Non-Travel  \
0                       2                        0.0   
1                       7                        0.0   
2                       0                        0.0   
3                       2                        0.0   
4                       2                        0.0   
..                    ...                        ...   
363                     3                        0.0   
364                     8                        0.0   
365                     8                        0.0   
366                  

In [15]:
X_train_final.columns

Index(['Age', 'DistanceFromHome', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears',
       'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany',
       'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager',
       'BusinessTravel_Non-Travel', 'BusinessTravel_Travel_Frequently',
       'BusinessTravel_Travel_Rarely', 'OverTime_Yes'],
      dtype='object')

In [16]:
# Create a StandardScaler
X_scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler.fit(X_train_final)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train_final)
X_test_scaled = X_scaler.transform(X_test_final)


In [17]:
# Create a OneHotEncoder for the Department column
y_encoder_dept = OneHotEncoder(sparse=False)

# Fit the encoder to the training data
# Create two new variables by applying the encoder
# to the training and testing data

y_encoded_dept_train = y_encoder_dept.fit_transform(y_train[['Department']])
y_encoded_dept_test = y_encoder_dept.fit_transform(y_test[['Department']])

print(y_encoded_dept_train[0:5])
print(y_encoded_dept_test[0:5])


[[0. 1. 0.]
 [0. 0. 1.]
 [0. 0. 1.]
 [0. 0. 1.]
 [0. 0. 1.]]
[[0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [18]:
# Create a OneHotEncoder for the Attrition column
y_encoder_attr = OneHotEncoder(sparse=False)

# Fit the encoder to the training data
# Create two new variables by applying the encoder
# to the training and testing data

y_encoded_attr_train = y_encoder_dept.fit_transform(y_train[['Attrition']])
y_encoded_attr_test = y_encoder_dept.fit_transform(y_test[['Attrition']])

print(y_encoded_attr_train[0:5])
print(y_encoded_attr_test[0:5])


[[0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]
[[1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]]


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


## Create, Compile, and Train the Model

In [19]:
# Find the number of columns in the X training data
input_nodes = len(X_train_final.columns)

# Create the input layer
input_layer = layers.Input(shape=(input_nodes,), name='input_layer')

# Create at least two shared layers. Started with 24 neurons, because that is the number of inputs. Then tried again with 48.
shared_layer1 = layers.Dense(48, activation='relu', name='shared_layer1')(input_layer)
shared_layer2 = layers.Dense(24, activation= 'relu', name='shared_layer2')(shared_layer1)


In [20]:
# Create a branch for Department
# with a hidden layer and an output layer


# Create the hidden layer
dept_hidden = Dense(6, activation='relu', name='dept_hidden')(shared_layer2)

# Create the output layer. Using softmax because there are 3 possible departments and want to know the probability in relation to each other.
dept_output = Dense(3, activation='softmax', name='dept_output')(dept_hidden)


In [21]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attr_hidden = Dense(6, activation='relu', name='attr_hidden')(shared_layer2)

# Create the output layer
attr_output = Dense(2, activation='sigmoid', name='attr_output')(attr_hidden)


In [22]:
# Create the model
model = Model(inputs=input_layer, outputs=[dept_output, attr_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'dept_output': 'categorical_crossentropy',
                    'attr_output': 'binary_crossentropy'},
              metrics={'dept_output': 'accuracy',
                       'attr_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_layer (InputLayer)    [(None, 24)]                 0         []                            
                                                                                                  
 shared_layer1 (Dense)       (None, 48)                   1200      ['input_layer[0][0]']         
                                                                                                  
 shared_layer2 (Dense)       (None, 24)                   1176      ['shared_layer1[0][0]']       
                                                                                                  
 dept_hidden (Dense)         (None, 6)                    150       ['shared_layer2[0][0]']       
                                                                                              

In [23]:
# Train the model
model.fit(
    X_train_final,
    {'dept_output': y_encoded_dept_train, 'attr_output': y_encoded_attr_train},
    epochs = 20,
    batch_size = 32
)



Epoch 1/20
35/35 [==============================] - 4s 10ms/step - loss: 2.5314 - dept_output_loss: 1.9755 - attr_output_loss: 0.5559 - dept_output_accuracy: 0.5381 - attr_output_accuracy: 0.8367
Epoch 2/20
35/35 [==============================] - 0s 9ms/step - loss: 1.3101 - dept_output_loss: 0.8511 - attr_output_loss: 0.4590 - dept_output_accuracy: 0.6425 - attr_output_accuracy: 0.8448
Epoch 3/20
35/35 [==============================] - 1s 16ms/step - loss: 1.1522 - dept_output_loss: 0.7160 - attr_output_loss: 0.4363 - dept_output_accuracy: 0.7123 - attr_output_accuracy: 0.8439
Epoch 4/20
35/35 [==============================] - 0s 8ms/step - loss: 1.1130 - dept_output_loss: 0.6854 - attr_output_loss: 0.4276 - dept_output_accuracy: 0.7278 - attr_output_accuracy: 0.8457
Epoch 5/20
35/35 [==============================] - 0s 5ms/step - loss: 1.0830 - dept_output_loss: 0.6724 - attr_output_loss: 0.4105 - dept_output_accuracy: 0.7432 - attr_output_accuracy: 0.8430
Epoch 6/20
35/35 [=====

In [24]:
# Evaluate the model with the testing data
results = model.evaluate(np.array(X_test_final), {
    'dept_output': y_encoded_dept_test,
    'attr_output': y_encoded_attr_test
})

12/12 [==============================] - 0s 7ms/step - loss: 1.0818 - dept_output_loss: 0.6431 - attr_output_loss: 0.4387 - dept_output_accuracy: 0.8261 - attr_output_accuracy: 0.8261


In [26]:
# Print the accuracy for both department and attrition
pred_categories = ['Attrition', 'Department']
for i, cat in enumerate(pred_categories):
    print(f"{cat} accuracy: {results[i+3]}")

Attrition accuracy: 0.8260869383811951
Department accuracy: 0.8260869383811951


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

In [38]:
attrition_df.value_counts('Attrition')

Attrition
No     1233
Yes     237
dtype: int64

YOUR ANSWERS HERE

1. For **Attrition**, accuracy may not be the best metric because the original data is not balanced between 'yes' and 'no' responses, as can be seen in the cell directly above (No = 1233, Yes = 237). F1 score would likely give a better understanding of the model's efficacy as it balances precision and recall.
For **Department**, this is a categorical prediction with three possible responses. While the accuracy score gives an overall sense of how often the model is correct, a weighted-average F1 score and a confusion matrix would likely be more helpful to understand where the model is making mistakes.
2. For **Attrition** I used sigmoid for the activation function because it is a binary output. For **Department** I used softmax for the activation function because there are 3 possible outcomes.
3. I think it would be more accurate for each branch to actually have separate models for each. However, that requires more work, and is less efficient. I also think additional metrics should be used to evaluate the model, as mentioned in number 1 above.  Additionally, I did not use all the columns originally provided, I removed some.  It's possible that including some of the columns I dropped, or removing some other columns would make the model more effective. It would also be possible to use some optimization tools to understand what the optimal parameters are for the model, as the epochs and nodes I used were just a best guess.